server/jobs/job_manager.go

```diff
@@ -0,0 +1,143 @@
+package jobs
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lieut-data/mattermost-plugin-api/cluster"
+	"github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar"
+	"github.com/pkg/errors"
+)
+
+type JobManager struct {
+	registeredJobs sync.Map
+	activeJobs     sync.Map
+	env            mscalendar.Env
+	papi           cluster.JobPluginAPI
+}
+
+type RegisteredJob struct {
+	id                string
+	interval          time.Duration
+	work              func(env mscalendar.Env)
+	isEnabledByConfig func(env mscalendar.Env) bool
+}
+
+type activeJob struct {
+	RegisteredJob
+
+	cancel     chan struct{}
+	cancelled  chan struct{}
+	cancelOnce sync.Once
+}
+
+// NewJobManager creates a JobManager for to let plugin.go coordinate with the scheduled jobs.
+func NewJobManager(papi cluster.JobPluginAPI, env mscalendar.Env) *JobManager {
+	return &JobManager{
+		papi: papi,
+		env:  env,
+	}
+}
+
+// AddJob accepts a RegisteredJob, stores it, and activates it if enabled.
+func (jm *JobManager) AddJob(job RegisteredJob) error {
+	jm.registeredJobs.Store(job.id, job)
+	if job.isEnabledByConfig(jm.env) {
+		err := jm.activateJob(job)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// OnConfigurationChange activates/deactivates jobs based on their current state, and the current plugin config.
+func (jm *JobManager) OnConfigurationChange(env mscalendar.Env) error {
+	jm.env = env
+	jm.registeredJobs.Range(func(k interface{}, v interface{}) bool {
+		job, ok := v.(RegisteredJob)
+		if !ok {
+			jm.env.Logger.Errorf("Expected RegisteredJob, got %T", v)
+			return true
+		}
+		enabled := job.isEnabledByConfig(env)
+		active := jm.isJobActive(job.id)
+
+		// Config is set to enable. Job does not exist, create new job.
```

Comment:
nit: Regarding the comment here, the job exist, but it is not active right? so it would not create the job, just activate it.

Reply:
`Create` in this context means `activate` I suppose. I can change to say "job not active, so activate the job" or something like that.

Reply:
Sounds nice